### DISCUSSION TITLE
What is the best way to handle continuation packets?

### DISCUSSION THREAD

**Discussion by joeymukherjee** [CCSDSPy/ccsdspy#122] [Q&A]
I have packets making use of the CCSDS_SEQUENCE_FLAG and I need to put the packets together before using the rest of my parsing.  What is the best way to do this?

I have to use split_packet_bytes or iter_packet_bytes so deal with the packets at a byte level, right?

Best practices welcome!

**Comment by ddasilva** [comment]
Hello @joeymukherjee 

How do continuation packets combine in your case? Does a certain data field accumulate together?

**Reply by joeymukherjee** [reply]
Yes, I have a "DATA" field which gets appended to for subsequent packets.  We also have a "BYTE_COUNT" field which has the total amount for the DATA field in the first packet.  This is a brute force way of how I wrote it in code.  Not very efficient, but it does work for my small test case.  I'm not real sure how to do it better though.  Thank you for any insights!

          ```
    # full bytes is the entire files as bytes, but I save the start of the packets index in an array

    full_bytes = bytearray ()
    full_bytes_index = []
    for x in iter_packet_bytes (stream, include_primary_header=True):
        full_bytes_index.append (len (full_bytes))
        full_bytes.extend (x)

    packet_data = all_packet_types [mnemonic].load (io.BytesIO (full_bytes), include_primary_header=True)

     #  I need to copy the first part of the data array, but the packet_data ['DATA'][start_packet] is the right size already but filled with junk

    new_data_array = packet_data ['DATA'][start_packet][0:packet_data ['CCSDS_PACKET_LENGTH'][start_packet] - 25]  # where did 25 come from?

    # We have to put the continuation packets together

    start_packet = 0
    last_packet = 0
    for i, flag in enumerate (packet_data ['CCSDS_SEQUENCE_FLAG']):
       if flag == 3:   # unsegmented
          continue
       elif flag == 0: # continuation segment
          continue
       elif flag == 1: # first segment
          start_packet = i
          continue
       elif flag == 2: # last segment
          last_packet = i
          for x in range (start_packet + 1, last_packet + 1):
              start_pos = full_bytes_index [x] + 10                     # 10 is the start byte of where data is
              length = packet_data ['CCSDS_PACKET_LENGTH'][x] - 5     # 5 is the up to the CCSDS packet length - 1
              new_data_array = numpy.concatenate ((new_data_array, full_bytes [start_pos:start_pos + length]))
              for key in packet_data:  # delete all the entries for everything by the first packet!
                  packet_data [key] = numpy.delete (packet_data [key], x)
          assert len (new_data_array) == packet_data ['BYTE_COUNT'], f"DATA array is {len (new_data_array)} bytes long instead of {packet_data ['BYTE_COUNT'][start_packet]}"
```

**Reply by ddasilva** [reply]
This is what I would do too. Minor difference in what I would do would be that instead of keeping track of `start_packet` and `last_packet`, I would recommend you keep a buffer. With this buffer, you would:
- reset to empty when you see a start packet
- append to in a continuation segment
- finalize/process in a last segment

I think this pattern is a bit more standard, though both approaches are acceptable.

**Reply by ddasilva** [reply]
Did this help @joeymukherjee ? Let me know if there is anything else I can do

**Reply by joeymukherjee** [reply]
I'm not a fan of the buffer since appending to those are slower than keeping the segment counters, and for a large file, it might be significant.  That was one thing I didn't like about my implementation either: my `full_bytes` is a buffer and I am reading the entire file into memory and keeping the indexes of the start of the packets.  I also didn't like the fact that I am deleting the packets after it took time to parse them out.  It seems like I should be able to do that before the parsing process.

Could there be a way to an array of dictionaries instead of a dictionary of parallel arrays?  In some cases, this seems like it could be more useful?

However my code was working, so I didn't change it.   :) 

Thanks for the help though!
Joey

**Reply by ddasilva** [reply]
if you were thinking about the concatenation being slow from increasing copy overhead, it's not slower if you append each chunk to a list and do the concatenation before you finalize, or use list.extend or BytesIO. it's still O(n). what you're saying is true if you use a string or bytes object and do += each iteration (that is O(n^2))

**Reply by ddasilva** [reply]
> Could there be a way to an array of dictionaries instead of a dictionary of parallel arrays? In some cases, this seems like it could be more useful?

That's a good idea to reflect on; we've been doing it this way because it was originally easier to plot time series of housekeeping values. I'll give this suggestion some thought
